File: python/algo/202510/algo3134.py
from bisect import bisect_left
from collections import defaultdict
from typing import List

# 滑动窗口与双指针 - 求子数组个数 2.3.1 越短越合法
class Solution:
    def medianOfUniquenessArray(self, nums: List[int]) -> int:
        n = len(nums)
        k = (((n + 1) * n // 2) + 1) // 2 # ceil((n*(n+1)/2)/2)

        # upper 说明 upper 大于期望值，缩小
        def check(upper : int) -> int:
            freq = defaultdict(int)
            l = cnt = 0
            for r, v in enumerate(nums):
                freq[v] += 1
                while len(freq) > upper:
                    freq[nums[l]] -= 1
                    if (freq[nums[l]]) == 0: del freq[nums[l]]
                    l += 1
                cnt += r - l  + 1
                if cnt >= k:return True
            return False
        
        return bisect_left(range(len(set(nums))), True, 1, key=check)

File: cplus/leetcode/202510/algo3134.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 求子数组个数 2.3.1 越短越合法
class Solution {
public:
    int medianOfUniquenessArray(vector<int>& nums) {
        // 二分 + 滑动窗口
        int n = nums.size(); 
        long long k = ((1LL * n * (n + 1) / 2) + 1) / 2;

        auto check = [&](int upper) -> bool {
            int l = 0;
            long long cnt = 0;
            unordered_map<int, int> freq;
            for (int r=0; r<n; ++r) {
                freq[nums[r]] ++;
                while (freq.size() > upper) {
                    freq[nums[l]] --;
                    if (freq[nums[l]] == 0) freq.erase(nums[l]);
                    l += 1;
                }
                cnt += r - l + 1;
                if (cnt >= k) return true;  // 注意条件
            }
            return false;
        };

        set<int> distinct(nums.begin(), nums.end());
        int l = 0, r = distinct.size();
        while (l+1 < r) {
            int mid = (l + r) / 2;
            (check(mid)? r:l) = mid;
        }
        return r;

    }
};